---
title: "Dada2_pipeline_CIPF_2018"
author: "Lola Giner PÃ©rez"
date: "2023-03-28"
output: html_document
---

```{r Dada2 installation}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("dada2")
```

```{r Dada2 run}
library(dada2); packageVersion("dada2")
path <- "/fast/AG_Forslund/Lola/CIPF_2018/wetransfer_muestras-rungen22_2023-03-21_1356" 
list.files(path)


# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_R"), `[`, 1)

plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:2])

# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))


out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(250,210),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, trimLeft = c(30, 30),
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)


errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)

derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
names(derepRs) <- sample.names

dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
dadaFs[[1]]

mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])

seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
#Reduce noise
seqtab2 <- seqtab[,nchar(colnames(seqtab)) %in% seq(379,386)]

seqtab.nochim <- removeBimeraDenovo(seqtab2, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab2)

getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)

colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)

##Create an ASV matrix with ASV as rows and samples as columns
asvmat <- t(seqtab.nochim) #Removing sequence rownames for display only
asv.names <- as.data.frame(rownames(asvmat))
rownames(asv.names) <- paste0("ASV", 1:nrow(asv.names))
rownames(asvmat)<-NULL
rownames(asvmat) <- paste0("ASV", 1:nrow(asvmat))
#colnames(asvmat) <- paste0("Sample", 1:ncol(asvmat)) #keep original sample ID
head(asvmat)
write.csv(asvmat, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/output/Tables/otu_table.csv", row.names = F, col.names = F)

##Get count of ASVs detected by sample
asv.sample<- as.data.frame(asvmat)
test<- data.frame()
for (i in 1:ncol(asv.sample)) {
    asv<- data.frame()
    asv[1,1]<- sum(asv.sample[,i]!=0)
    rownames(asv)<- paste0("Sample", i)
    test <- rbind(test, asv) ### Join all the "individual" data frames into the final data frame 
}
asv.sample<- as.matrix(test)
colnames(asv.sample)<- "ASVs_dada2"
rm(test,asv, i)
#Important not to loose the connection between the seq and the ASVs
write.csv(asv.names, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/output/Tables/asv_table.csv", row.names = F, col.names = F)

###Track reads through the pipeline 
getN <- function(x) sum(getUniques(x))
track <- cbind(sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(filtFs, getN), sapply(filtRs, getN), sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
#If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
track<-track[,c(1,3,5:8)]
colnames(track) <- c("input_dada2", "filtered_dada2", "denoisedF_dada2", "denoisedR_dada2", "merged_dada2", "nonchim_dada2")
rownames(track) <- sample.names
rownames(asv.sample) <- sample.names
track<-cbind(track, asv.sample)
head(track)

  

taxa <- assignTaxonomy(seqtab.nochim, "/fast/AG_Forslund/Lola/SILVA_databases/SILVA_db", multithread=TRUE)

taxa <- addSpecies(taxa, "/fast/AG_Forslund/Lola/SILVA_databases/silva_species_assignment_v138.1.fa.gz")

taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)

write.csv(taxa, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/output/Tables/tax_table.csv", row.names = F, col.names = F)


library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
theme_set(theme_bw())
samples.out <- rownames(seqtab.nochim)


ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(), 
               tax_table(taxa))
ps <- prune_samples(sample_names(ps) != "Mock", ps) # Remove mock sample
ps
