---
title: "Dada2_pipeline_CIPF_2018"
author: "Lola Giner PÃ©rez"
date: "2023-03-28"
output: html_document
---

```{r Dada2 installation}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("dada2")
```


```{r Dada2 run}
library(dada2); packageVersion("dada2")
path <- "/fast/AG_Forslund/Lola/CIPF_2018/wetransfer_muestras-rungen22_2023-03-21_1356" 
list.files(path)


# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_R"), `[`, 1)

plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:2])

# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))


out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(250,210),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, trimLeft = c(30, 30),
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)


errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)

derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
names(derepRs) <- sample.names

dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
dadaFs[[1]]

mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])

seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
#Reduce noise
seqtab2 <- seqtab[,nchar(colnames(seqtab)) %in% seq(379,386)]

seqtab.nochim <- removeBimeraDenovo(seqtab2, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab2)

getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)

colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)

##Create an ASV matrix with ASV as rows and samples as columns
asvmat <- t(seqtab.nochim) #Removing sequence rownames for display only
asv.names <- as.data.frame(rownames(asvmat))
rownames(asv.names) <- paste0("ASV", 1:nrow(asv.names))
rownames(asvmat)<-NULL
asvmat <- as.data.frame(asvmat)
rownames(asvmat) <- paste0("ASV", 1:nrow(asvmat))
#colnames(asvmat) <- paste0("Sample", 1:ncol(asvmat)) #keep original sample ID
head(asvmat)
write.csv(asvmat, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/otu_table_silva2.csv", row.names = F, col.names = F)

#Creating the OTU table saving the ASV names
make_csv_table(asvmat, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/otu_table_silva2", info = "otu_table")

##Get count of ASVs detected by sample
asv.sample<- as.data.frame(asvmat)
test<- data.frame()
for (i in 1:ncol(asv.sample)) {
    asv<- data.frame()
    asv[1,1]<- sum(asv.sample[,i]!=0)
    rownames(asv)<- paste0("Sample", i)
    test <- rbind(test, asv) ### Join all the "individual" data frames into the final data frame 
}
asv.sample<- as.matrix(test)
colnames(asv.sample)<- "ASVs_dada2"
rm(test,asv, i)
#Important not to loose the connection between the seq and the ASVs
write.csv(asv.names, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/asv_table_silva2.csv", row.names = F, col.names = F)

###Track reads through the pipeline 
getN <- function(x) sum(getUniques(x))
track <- cbind(sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(filtFs, getN), sapply(filtRs, getN), sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
#If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
track<-track[,c(1,3,5:8)]
colnames(track) <- c("input_dada2", "filtered_dada2", "denoisedF_dada2", "denoisedR_dada2", "merged_dada2", "nonchim_dada2")
rownames(track) <- sample.names
rownames(asv.sample) <- sample.names
track<-cbind(track, asv.sample)
head(track)

  

taxa <- assignTaxonomy(seqtab.nochim, "/fast/AG_Forslund/Lola/SILVA_databases/SILVA2/silva_nr99_v138.1_train_set.fa.gz", multithread=TRUE)

taxa <- addSpecies(taxa, "/fast/AG_Forslund/Lola/SILVA_databases/SILVA2/silva_species_assignment_v138.1.fa.gz")

taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)

write.csv(taxa, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/tax_table_with_seq_silva2.csv", row.names = F, col.names = F)


##Create an Taxa matrix with ASV as rows and taxonomic level as columns
taxa_asvs <- as.data.frame(taxa) # Removing sequence rownames for display only
rownames(taxa_asvs) <- NULL
rownames(taxa_asvs) <- paste0("ASV", 1:nrow(asv.names))
head(taxa_asvs)
make_csv_table(taxa_asvs, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/tax_table_silva", info = "tax_table")


##Transform seqtab.nochim to fasta file 
  library(DECIPHER); packageVersion("DECIPHER")
  ##Create a DNAString set from the ASVs
  dna <- DNAStringSet(getSequences(seqtab.nochim))
  names(dna)<- paste0("ASV", 1:length(dna)) ##Give short names to each sequence
 writeXStringSet(dna, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/ASV_silva2.fasta") ##Export fasta seq
  
  ##Create a Phylogenetic tree
  library(phangorn)
  Align16S<- AlignSeqs(dna, anchor= NA, verbose= FALSE) ##Alignment
  
  phangAlign16S <- phyDat(as(Align16S, "matrix"), type="DNA")
  dm16S <- dist.ml(phangAlign16S) ## Distance matrix
  treeNJ16S <- NJ(dm16S) # Note, tip order != sequence order
  plot(treeNJ16S, type= "unrooted", use.edge.length= FALSE, no.margin= TRUE, show.tip.label= TRUE) ##Neighbor-Joining tree
  fit1 <- pml(treeNJ16S, data=phangAlign16S)
  fitGTR16S <- update(fit1, k=4, inv=0.2)
  fitGTR16S <- optim.pml(fitGTR16S, model="GTR", optInv=TRUE, optGamma=TRUE,
                         rearrangement = "stochastic", control = pml.control(trace = 0))
  plot(fitGTR16S, type= "unrooted", use.edge.length= FALSE, no.margin= TRUE, show.tip.label= FALSE)
  
  tree<- fitGTR16S$tree

saveRDS(fitGTR16S, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/phylo_tree_silva2.rds")  
  
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
theme_set(theme_bw())
samples.out <- rownames(seqtab.nochim)

```



```{r assign taxonomy RDP}
taxa <- assignTaxonomy(seqtab.nochim, "/fast/AG_Forslund/Lola/SILVA_databases/RDP_database/rdp_train_set_18.fa.gz", multithread=TRUE)

taxa <- addSpecies(taxa, "/fast/AG_Forslund/Lola/SILVA_databases/RDP_database/rdp_species_assignment_18.fa.gz")

taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)

write.csv(taxa, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/tax_table_with_seq_rdp.csv", row.names = F, col.names = F)


##Create an Taxa matrix with ASV as rows and taxonomic level as columns
taxa_asvs <- as.data.frame(taxa) # Removing sequence rownames for display only
rownames(taxa_asvs) <- NULL
rownames(taxa_asvs) <- paste0("ASV", 1:nrow(asv.names))
head(taxa_asvs)
make_csv_table(taxa_asvs, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/tax_table_rdp", info = "tax_table")


##Transform seqtab.nochim to fasta file 
  library(DECIPHER); packageVersion("DECIPHER")
  ##Create a DNAString set from the ASVs
  dna <- DNAStringSet(getSequences(seqtab.nochim))
  names(dna)<- paste0("ASV", 1:length(dna)) ##Give short names to each sequence
 writeXStringSet(dna, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/ASV_rdp.fasta") ##Export fasta seq
  
  ##Create a Phylogenetic tree
  library(phangorn)
  Align16S<- AlignSeqs(dna, anchor= NA, verbose= FALSE) ##Alignment
  
  phangAlign16S <- phyDat(as(Align16S, "matrix"), type="DNA")
  dm16S <- dist.ml(phangAlign16S) ## Distance matrix
  treeNJ16S <- NJ(dm16S) # Note, tip order != sequence order
  plot(treeNJ16S, type= "unrooted", use.edge.length= FALSE, no.margin= TRUE, show.tip.label= TRUE) ##Neighbor-Joining tree
  fit1 <- pml(treeNJ16S, data=phangAlign16S)
  fitGTR16S <- update(fit1, k=4, inv=0.2)
  fitGTR16S <- optim.pml(fitGTR16S, model="GTR", optInv=TRUE, optGamma=TRUE,
                         rearrangement = "stochastic", control = pml.control(trace = 0))
  plot(fitGTR16S, type= "unrooted", use.edge.length= FALSE, no.margin= TRUE, show.tip.label= FALSE)
  
  tree<- fitGTR16S$tree

saveRDS(fitGTR16S, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/phylo_tree_rdp.rds")  
  
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
theme_set(theme_bw())
samples.out <- rownames(seqtab.nochim)

```

```{r phyloseq silva}
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
library(readr)
library(readxl)
library(readr)
library(vegan)
library(tidyverse)


otu_table <- read_delim("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/otu_table_silva2.csv", 
    delim = ";", escape_double = FALSE, trim_ws = TRUE)
tax_table <- read_delim("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/tax_table.csv", 
    delim = ";", escape_double = FALSE, trim_ws = TRUE)

metadata <- read_excel("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/metadata.xlsx")
tree<- readRDS("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/phylo_tree_silva2.rds")
dna<- readDNAStringSet("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/ASV_silva2.fasta")


otu_table <- as.data.frame(otu_table)
rownames(otu_table) <- otu_table$...1
otu_ok <- subset(otu_table, select = -c(CV1_S1, CV16_S2, CR23_S3, CR24_S4, CC34_36_S5, CC37_1_S6, CCR55_S7, CCR56_S8, ...1))
otu_ok <- t(otu_ok)%>%subset(select = -c(2628))%>%t()%>%as.data.frame()


tax_table <- as.data.frame(tax_table)
rownames(tax_table) <- tax_table$...1
tax_table<-subset(tax_table, select=-c(`...1`))



 OTU = otu_table(otu_ok, taxa_are_rows = TRUE)
 TAX = tax_table(as.matrix(tax_table))
 SAMPLE = sample_data(metadata)
 TREE = tree<- tree$tree
 
rownames(SAMPLE) <-metadata$SampleID

 ps_silva<- phyloseq(OTU, TAX, SAMPLE, TREE)
 ps_silva
saveRDS(ps_silva, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/ps_silva.Rds")  

```

```{r phyloseq rdp}
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
library(readr)
library(readxl)
library(readr)
library(vegan)
library(tidyverse)

#La otu table se crea antes que asignarse la taxonomia (asvmat) por eso usamos la misma
otu_table <- read_delim("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/otu_table_silva2.csv", 
    delim = ";", escape_double = FALSE, trim_ws = TRUE)
tax_table <- read_delim("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/Tables/tax_table_rdp.csv", 
    delim = ";", escape_double = FALSE, trim_ws = TRUE)

metadata <- read_excel("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/metadata.xlsx")
tree<- readRDS("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/phylo_tree_rdp.rds")
dna<- readDNAStringSet("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/ASV_rdp.fasta")


otu_table <- as.data.frame(otu_table)
rownames(otu_table) <- otu_table$...1
otu_rdp <- subset(otu_table, select = -c(CV1_S1, CV16_S2, CR23_S3, CR24_S4, CC34_36_S5, CC37_1_S6, CCR55_S7, CCR56_S8, ...1))
otu_rdp <- t(otu_rdp)%>%subset(select = -c(2628))%>%t()%>%as.data.frame()


tax_table <- as.data.frame(tax_table)
rownames(tax_table) <- tax_table$...1
tax_table<-subset(tax_table, select=-c(`...1`))

 OTU = otu_table(otu_rdp, taxa_are_rows = TRUE)
 TAX = tax_table(as.matrix(tax_table))
 SAMPLE = sample_data(metadata)
 TREE = tree<- tree$tree
 
rownames(SAMPLE) <-metadata$SampleID
 ps_rdp<- phyloseq(OTU, TAX, SAMPLE, TREE)
 ps_rdp
saveRDS(ps_rdp, "/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/ps_rdp.Rds")
```

#Barplots phylum SILVA no filters
```{r Barplots Phylum}
ps_silva<-readRDS("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/ps_silva.Rds")
ps.rel = transform_sample_counts(ps_silva, function(x) x/sum(x)*100)
# agglomerate taxa
glom <- tax_glom(ps.rel, taxrank = 'Phylum', NArm = FALSE)
ps.melt <- psmelt(glom)
# change to character for easy-adjusted level
ps.melt$Phylum <- as.character(ps.melt$Phylum)
ps.melt <- ps.melt %>%
    group_by(Group, Phylum) %>%
    mutate(median=median(Abundance))

#to get the same rows together
ps.melt_sum <- ps.melt %>%
    group_by(Sample,Group,Phylum) %>%
    summarise(Abundance=sum(Abundance))%>%
    arrange(Group) %>%
    mutate(Group = factor(Group, levels=c("ctr", "ctr+rif", "ccl4", "ccl4+rif")))

ggplot(ps.melt_sum, aes(x = Sample, y = Abundance, fill = Phylum)) + 
    geom_bar(stat = "identity", aes(fill=Phylum)) + 
    labs(x="", y="%") +
    facet_wrap(~Group, scales= "free_x", nrow=1) +
    theme_classic() + 
    theme(strip.background = element_blank(), 
          axis.text.x.bottom = element_text(angle = -90))
```

#Barplots Phylum RDP no filters
```{r Barplots Phylum}
ps_rdp<-readRDS("/fast/AG_Forslund/Lola/CIPF_2018/Analysis/Rats_HE_CCL4model/output2/ps_rdp.Rds")
ps.rel = transform_sample_counts(ps_rdp, function(x) x/sum(x)*100)
# agglomerate taxa
glom <- tax_glom(ps.rel, taxrank = 'Phylum', NArm = FALSE)
ps.melt <- psmelt(glom)
# change to character for easy-adjusted level
ps.melt$Phylum <- as.character(ps.melt$Phylum)
ps.melt <- ps.melt %>%
    group_by(Group, Phylum) %>%
    mutate(median=median(Abundance))

#to get the same rows together
ps.melt_sum <- ps.melt %>%
    group_by(Sample,Group,Phylum) %>%
    summarise(Abundance=sum(Abundance))%>%
    arrange(Group) %>%
    mutate(Group = factor(Group, levels=c("ctr", "ctr+rif", "ccl4", "ccl4+rif")))

ggplot(ps.melt_sum, aes(x = Sample, y = Abundance, fill = Phylum)) + 
    geom_bar(stat = "identity", aes(fill=Phylum)) + 
    labs(x="", y="%") +
    facet_wrap(~Group, scales= "free_x", nrow=1) +
    theme_classic() + 
    theme(strip.background = element_blank(), 
          axis.text.x.bottom = element_text(angle = -90))
```